File: py/2261.py
from typing import List, Iterable
import collections


class hash_list(list):
    def __init__(self, *args):
        if len(args) == 1 and isinstance(args[0], Iterable):
            args = args[0]
        super().__init__(args)

    def __hash__(self):
        return hash(e for e in self)

    def __eq__(self, other):
        return super().__eq__(other)


class Range:
    def __init__(self, nums, begin, end):
        self.nums = nums
        self.begin = begin
        self.end = end

    def __hash__(self):
        return hash(tuple(self.nums[self.begin:self.end]))

    def __eq__(self, other):
        if self.end - self.begin != other.end - other.begin:
            return False
        return self.nums[self.begin:self.end] == other.nums[other.begin:other.end]

    def __repr__(self) -> str:
        return f"{self.nums[self.begin:self.end]}"


class Solution:
    def countDistinct(self, nums: List[int], k: int, p: int) -> int:
        div_pos = {x for x in range(len(nums)) if nums[x] % p == 0}

        unique_sets = collections.defaultdict(lambda: [])

        for begin_idx in range(0, len(nums)):
            for end_idx in range(begin_idx+1, len(nums)+1):
                in_range_div = [x for x in div_pos if begin_idx <= x < end_idx]
                if len(in_range_div) > k:
                    break

                this_range = Range(nums, begin_idx, end_idx)
                this_hash = hash(this_range)
                existing = unique_sets[this_hash]
                if this_range in existing:
                    continue
                existing.append(this_range)

        return sum((len(value) for value in unique_sets.values()))


numsa = [2, 3, 3, 2, 2]

r1 = Range(numsa, 0, 1)
r2 = Range(numsa, 3, 4)

s = Solution()
print(s.countDistinct(numsa, 2, 2))

numsa = [1, 2, 3, 4]
print(s.countDistinct(numsa, 4, 1))

File: cpp/2200_2299/2261.cpp
#include <vector>
#include <cassert>
#include <iostream>
#include <set>
#include <functional>
#include <map>
using namespace std;

struct Range{
    int begin;
    int end;
};

template<typename T> void hash_combine(size_t & seed, T const& v){
    std::hash<T> hasher;
    seed ^= hasher(v) + 0x9e3779b9 + (seed<<6) + (seed>>2);
}

template<typename It> std::size_t hash_range(It first, It last){
    size_t seed = 0;
    for(; first != last; ++first){
        hash_combine(seed, *first);
    }
    return seed;
}

void print(vector<int>& nums, multimap<size_t, Range> ranges){
    for(auto &kv: ranges){

        for(int i = kv.second.begin; i < kv.second.end; ++i){
            cout<<nums[i]<<",";
        }
        cout<<endl;
    }
}

class Solution {
public:
    int countDistinct(vector<int>& nums, int k, int p) {
        std::set<int> div_pos;
        for (int i=0; i< nums.size();++i){
            if(nums[i] % p == 0){
                div_pos.insert(i);
            }
        }
        //           2,3,3,2,2
        // div_pos = 0     3,4
        // 

        std::multimap<size_t, Range> unique_sets;
        for(int begin_idx = 0; begin_idx < nums.size() ; ++begin_idx){
            for(int end_idx = begin_idx + 1; end_idx <= nums.size(); ++ end_idx){
                // find how many div_pos in [begin_idx, end_idx)

                auto div_begin = div_pos.upper_bound(begin_idx-1);
                auto div_end = div_pos.lower_bound(end_idx);

                //auto div_begin_offset = distance(div_pos.begin(), div_begin);
                //auto div_end_offset = distance(div_pos.begin(), div_end);

                int div_number = std::distance(div_begin, div_end);

                assert (div_number >= 0);
                assert (div_number <= div_pos.size());

                if (div_number > k){
                    break;
                }

                Range this_range = {begin_idx, end_idx};

                auto this_hash = hash_range(nums.begin()+ begin_idx, nums.begin() + end_idx);

                auto existing_range = unique_sets.equal_range(this_hash);
               
                auto new_size = end_idx - begin_idx;

                bool equal_set = false;
                for (auto it = existing_range.first; it != existing_range.second; ++it){
                    auto old_size = it->second.end - it->second.begin;

                    if(old_size != new_size){
                        continue;
                    }

                    //size equal, check if range is equal
                    bool range_equal = std::equal(nums.begin() + begin_idx, nums.begin() + end_idx,
                                                  nums.begin() + it->second.begin);

                    if (range_equal){
                        equal_set = true;
                        break;
                    }
                }

                if(!equal_set){
                    unique_sets.insert({this_hash, this_range});
                }
            }
        }

        //print(nums, unique_sets);
        return unique_sets.size();
    }
};

int main(){
    Solution s;
    std::vector<int> nums = {2,3,3,2,2};
    cout<<s.countDistinct(nums, 2,2) <<endl;

    nums = {1,2,3,4};
    cout<<s.countDistinct(nums, 4,1) <<endl;
}


